### PR TITLE
Adds the Gossip CRDT 

### PR DESCRIPTION
**Pull request by lasarojc**
This PR adds the `query` functionality to all example CRDT.
Also adds the skeleton for the Gossip CRDT.

Some comments in #822 are addressed in this PR

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -387,7 +387,7 @@ func (cs *State) OnStart() error {
 }
 
 // timeoutRoutine: receive requests for timeouts on tickChan and fire timeouts on tockChan
-// receiveRoutine: serializes processing of proposoals, block parts, votes; coordinates state transitions
+// receiveRoutine: serializes processing of proposals, block parts, votes; coordinates state transitions
```
Comment (lasarojc) [inline consensus/state.go:390]:
```suggestion
// receiveRoutine: serializes processing of proposoals, block parts, votes; coordinates state transitions
```

File: consensus/state.go [hunk 2]
```diff
@@ -2001,7 +2001,7 @@ func (cs *State) handleCompleteProposal(blockHeight int64) {
 	if cs.Step <= cstypes.RoundStepPropose && cs.isProposalComplete() {
 		// Move onto the next step
 		cs.enterPrevote(blockHeight, cs.Round)
-		if hasTwoThirds { // this is optimisation as this will be triggered when prevote is added
+		if hasTwoThirds { // this is an optimization as this will be triggered when prevote is added
```
Comment (lasarojc) [inline consensus/state.go:2004]:
```suggestion
		if hasTwoThirds { // this is optimization as this will be triggered when prevote is added
```

**Comment by faddat** [comment]
This is very interesting. Is it still on the roadmap?

**Comment by melekes** [comment]
CRDT as a conflict-free replicated data type?

**Comment by faddat** [comment]
@melekes yeah so, to expand on what I found interesting -- I had not thought of gossip this way, but after thinking about it some CRDT does seem to encapsulate the ideal of what we'd do with gossip -- both block gossip and tx/mempool gossip, even. 

It should also be able to avoid a great deal of the noise that we are currently dealing with in gossip (I think).

**Comment by lasarojc** [comment]
We haven't allocated time for it but this is something I would very much like to see continued.

**Comment by faddat** [comment]
Yeah, I guess my thoughts/comment on this concept is that it seems elegant and efficent.

**Comment by lasarojc** [comment]
I am merging this into the base branch and opening a PR from there, so that the whole work is captured in a single place.

**Review by melekes — APPROVED** [review]
👍

**Comment by sergio-mena** [comment]
👍 here is is :-)

**Pull request merged**